Fails to build with FFmpeg 8

### Describe the bug

Fails to build

### What operating system(s) this bug have occurred on?

FreeBSD amd64

### What version(s) of DOSBox-X have this bug?

2025.05.03

### Output log

```text
<snip>
hardware.cpp:134:17: error: cannot initialize a parameter of type 'AVIOContext *' with an lvalue of type 'AVCodecContext *'
  134 |                 avcodec_close(ffmpeg_vid_ctx);
      |                               ^~~~~~~~~~~~~~
/usr/local/include/libavformat/avio.h:693:29: note: passing argument to parameter 's' here
  693 | int avio_close(AVIOContext *s);
      |                             ^
hardware.cpp:139:3: error: use of undeclared identifier 'avcodec_close'; did you mean 'avio_close'?
  139 |                 avcodec_close(ffmpeg_aud_ctx);
      |                 ^~~~~~~~~~~~~
      |                 avio_close
/usr/local/include/libavformat/avio.h:693:5: note: 'avio_close' declared here
  693 | int avio_close(AVIOContext *s);
      |     ^
hardware.cpp:139:17: error: cannot initialize a parameter of type 'AVIOContext *' with an lvalue of type 'AVCodecContext *'
  139 |                 avcodec_close(ffmpeg_aud_ctx);
      |                               ^~~~~~~~~~~~~~
/usr/local/include/libavformat/avio.h:693:29: note: passing argument to parameter 's' here
  693 | int avio_close(AVIOContext *s);
      |                             ^
hardware.cpp:174:20: error: no member named 'key_frame' in 'AVFrame'
  174 |         ffmpeg_aud_frame->key_frame = 1;
      |         ~~~~~~~~~~~~~~~~  ^
hardware.cpp:429:3: error: use of undeclared identifier 'avcodec_close'; did you mean 'avio_close'?
  429 |                 avcodec_close(ffmpeg_vid_ctx);
      |                 ^~~~~~~~~~~~~
      |                 avio_close
/usr/local/include/libavformat/avio.h:693:5: note: 'avio_close' declared here
  693 | int avio_close(AVIOContext *s);
      |     ^
hardware.cpp:429:17: error: cannot initialize a parameter of type 'AVIOContext *' with an lvalue of type 'AVCodecContext *'
  429 |                 avcodec_close(ffmpeg_vid_ctx);
      |                               ^~~~~~~~~~~~~~
/usr/local/include/libavformat/avio.h:693:29: note: passing argument to parameter 's' here
  693 | int avio_close(AVIOContext *s);
      |                             ^
hardware.cpp:1274:30: error: use of undeclared identifier 'FF_PROFILE_AAC_LOW'
 1274 |                         ffmpeg_aud_ctx->profile = FF_PROFILE_AAC_LOW;
      |                                                   ^
hardware.cpp:1534:23: error: no member named 'key_frame' in 'AVFrame'
 1534 |                                 ffmpeg_vid_frame->key_frame = ((capture.video.frames % 15) == 0)?1:0;
      |                                 ~~~~~~~~~~~~~~~~  ^
hardware.cpp:1771:1: warning: typedef requires a name [-Wmissing-declarations]
 1771 | typedef struct pcap_hdr_struct_t {
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hardware.cpp:1781:1: warning: typedef requires a name [-Wmissing-declarations]
 1781 | typedef struct pcaprec_hdr_struct_t {
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings and 13 errors generated.
```

Full log:

[dosbox-x-2025.05.03.log](https://github.com/user-attachments/files/21957007/dosbox-x-2025.05.03.log)

### Additional information

_No response_

### Have you checked that no similar bug report(s) exist?

- [x] I have searched and didn't find any similar bug report.

### Code of Conduct & Contributing Guidelines

- [x] I agree to follow the code of conduct and the contributing guidelines.

hardware.cpp:1530:17: error: use of undeclared identifier 'av_opt_set_int'; did you mean 'av_dict_set_int'?
 1530 |                 av_opt_set_int(ffmpeg_vid_ctx->priv_data, "g", 15, 0); // GOP size 15
      |                 ^~~~~~~~~~~~~~
      |                 av_dict_set_int

==> With original patch building against FFmpeg8:

```
<snip>
depbase=`echo pic.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`; c++ -DHAVE_CONFIG_H -I. -I../..   -I../../include -I../../src/aviwriter -I../../src/hardware/snd_pc98/sound -I../../src/hardware/snd_pc98/common -I../../sr
c/hardware/snd_pc98/generic -I../../src/hardware/snd_pc98/sound/getsnd -I../../src/hardware/snd_pc98/x11 -I../../src/hardware/snd_pc98/cbus -I../../src -Wno-int-to-void-pointer-cast   -Wno-address-of-packed-membe
r   -Wno-format-zero-length   -Wno-missing-field-initializers   -Wno-strict-aliasing   -Wno-implicit-fallthrough   -Wno-deprecated-declarations   -Wnon-virtual-dtor   -Wsuggest-override   -Wconversion-null   -Wsi
gn-promo   -Wno-error=incompatible-pointer-types   -Wno-error=format-security   -pedantic   -Wunused   -Wextra   -Wall  -DLIBICONV_PLUG -isystem /usr/local/include -I/usr/local/include -I/usr/local/include/SDL2 -
D_REENTRANT -D_THREAD_SAFE -I/usr/local/include -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03 -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/
vs/sdlnet/linux-host/include -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include/SDL -DLIBSLIRP_STATIC -I/usr/local/include/slirp -I/usr/local/include/gl
ib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include -I/usr/local/include  -pipe -DLIBICONV_PLUG -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -DLIBICONV_PLUG -isystem /usr/lo
cal/include -std=gnu++14  -O2 -msse  -Wall   -Wextra   -Wunused   -pedantic   -Wno-error=format-security   -Wno-error=incompatible-pointer-types   -Wsign-promo   -Wconversion-null   -Wsuggest-override   -Wnon-vir
tual-dtor   -Wno-deprecated-declarations   -Wno-implicit-fallthrough   -Wno-strict-aliasing   -Wno-missing-field-initializers   -Wno-format-zero-length   -Wno-address-of-packed-member   -Wno-int-to-void-pointer-c
ast  -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03 -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include -I/wrkdirs/ove
rlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include/SDL -MT pic.o -MD -MP -MF $depbase.Tpo -c -o pic.o pic.cpp && mv -f $depbase.Tpo $depbase.Po
--- hardware.o ---
hardware.cpp:1530:17: error: use of undeclared identifier 'av_opt_set_int'; did you mean 'av_dict_set_int'?
 1530 |                 av_opt_set_int(ffmpeg_vid_ctx->priv_data, "g", 15, 0); // GOP size 15
      |                 ^~~~~~~~~~~~~~
      |                 av_dict_set_int
/usr/local/include/libavutil/dict.h:174:5: note: 'av_dict_set_int' declared here
  174 | int av_dict_set_int(AVDictionary **pm, const char *key, int64_t value, int flags);
      |     ^
hardware.cpp:1530:32: error: cannot initialize a parameter of type 'AVDictionary **' with an lvalue of type 'void *'
 1530 |                 av_opt_set_int(ffmpeg_vid_ctx->priv_data, "g", 15, 0); // GOP size 15
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/libavutil/dict.h:174:36: note: passing argument to parameter 'pm' here
  174 | int av_dict_set_int(AVDictionary **pm, const char *key, int64_t value, int flags);
      |                                    ^
2 errors generated.
*** [hardware.o] Error code 1
```

==> With `av_opt_set_int(ffmpeg_vid_ctx->priv_data, "g", 15, 0); // GOP size 15` removed from patch:

```
<snip>
--- pci_bus.o ---
rk/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include/SDL -MT pci_bus.o -MD -MP -MF $depbase.
Tpo -c -o pci_bus.o pci_bus.cpp && mv -f $depbase.Tpo $depbase.Po
--- hardware.o ---
hardware.cpp:1530:23: error: no member named 'key_frame' in 'AVFrame'
 1530 |                                 ffmpeg_vid_frame->key_frame = ((capture.video.frames % 15) == 0)?1:0;
      |                                 ~~~~~~~~~~~~~~~~  ^
--- pic.o ---
depbase=`echo pic.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`; c++ -DHAVE_CONFIG_H -I. -I../..   -I../../include -I../../src/aviwriter -I../../src/hardware/snd_pc98/sound -I../../src/hardware/snd_pc98/common -I../../sr
c/hardware/snd_pc98/generic -I../../src/hardware/snd_pc98/sound/getsnd -I../../src/hardware/snd_pc98/x11 -I../../src/hardware/snd_pc98/cbus -I../../src -Wno-int-to-void-pointer-cast   -Wno-address-of-packed-membe
r   -Wno-format-zero-length   -Wno-missing-field-initializers   -Wno-strict-aliasing   -Wno-implicit-fallthrough   -Wno-deprecated-declarations   -Wnon-virtual-dtor   -Wsuggest-override   -Wconversion-null   -Wsi
gn-promo   -Wno-error=incompatible-pointer-types   -Wno-error=format-security   -pedantic   -Wunused   -Wextra   -Wall  -DLIBICONV_PLUG -isystem /usr/local/include -I/usr/local/include -I/usr/local/include/SDL2 -
D_REENTRANT -D_THREAD_SAFE -I/usr/local/include -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03 -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/
vs/sdlnet/linux-host/include -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include/SDL -DLIBSLIRP_STATIC -I/usr/local/include/slirp -I/usr/local/include/gl
ib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include -I/usr/local/include  -pipe -DLIBICONV_PLUG -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -DLIBICONV_PLUG -isystem /usr/lo
cal/include -std=gnu++14  -O2 -msse  -Wall   -Wextra   -Wunused   -pedantic   -Wno-error=format-security   -Wno-error=incompatible-pointer-types   -Wsign-promo   -Wconversion-null   -Wsuggest-override   -Wnon-vir
tual-dtor   -Wno-deprecated-declarations   -Wno-implicit-fallthrough   -Wno-strict-aliasing   -Wno-missing-field-initializers   -Wno-format-zero-length   -Wno-address-of-packed-member   -Wno-int-to-void-pointer-c
ast  -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03 -I/wrkdirs/overlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include -I/wrkdirs/ove
rlays/overlay/emulators/dosbox-x/work/dosbox-x-dosbox-x-v2025.05.03/vs/sdlnet/linux-host/include/SDL -MT pic.o -MD -MP -MF $depbase.Tpo -c -o pic.o pic.cpp && mv -f $depbase.Tpo $depbase.Po
--- hardware.o ---
1 error generated.
```

--- a/src/hardware/hardware.cpp
+++ b/src/hardware/hardware.cpp
@@ -1267,7 +1267,7 @@ skip_shot:
 			ffmpeg_aud_ctx->sample_rate = (int)capture.video.audiorate;
 			ffmpeg_aud_ctx->flags = 0; // do not use global headers
 			ffmpeg_aud_ctx->bit_rate = 320000;
-			ffmpeg_aud_ctx->profile = FF_PROFILE_AAC_LOW;
+			// ffmpeg_aud_ctx->profile = FF_PROFILE_AAC_LOW;
 
 			#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(59,24,100)
 			ffmpeg_aud_ctx->channels = 2;

--- a/src/hardware/hardware.cpp
+++ b/src/hardware/hardware.cpp
@@ -123,20 +123,18 @@ void ffmpeg_closeall() {
 			ffmpeg_avformat_began = false;
 		}
 		avio_close(ffmpeg_fmt_ctx->pb);
-		if (ffmpeg_vid_ctx != NULL) avcodec_close(ffmpeg_vid_ctx);
-		if (ffmpeg_aud_ctx != NULL) avcodec_close(ffmpeg_aud_ctx);
+		if (ffmpeg_vid_ctx != NULL) avcodec_free_context(&ffmpeg_vid_ctx);
+		if (ffmpeg_aud_ctx != NULL) avcodec_free_context(&ffmpeg_aud_ctx);
 		avformat_free_context(ffmpeg_fmt_ctx);
 		ffmpeg_fmt_ctx = NULL;
 		ffmpeg_vid_ctx = NULL; // NTS: avformat_free_context() freed this for us, don't free again
 		ffmpeg_aud_ctx = NULL; // NTS: avformat_free_context() freed this for us, don't free again
 	}
 	if (ffmpeg_vid_ctx != NULL) {
-		avcodec_close(ffmpeg_vid_ctx);
 		avcodec_free_context(&ffmpeg_vid_ctx);
 		ffmpeg_vid_ctx = NULL;
 	}
 	if (ffmpeg_aud_ctx != NULL) {
-		avcodec_close(ffmpeg_aud_ctx);
 		avcodec_free_context(&ffmpeg_aud_ctx);
 		ffmpeg_aud_ctx = NULL;
 	}
@@ -426,7 +424,6 @@ int ffmpeg_bpp_pick_rgb_format(int bpp) {
 
 void ffmpeg_reopen_video(double fps,const int bpp) {
 	if (ffmpeg_vid_ctx != NULL) {
-		avcodec_close(ffmpeg_vid_ctx);
 		avcodec_free_context(&ffmpeg_vid_ctx);
 		ffmpeg_vid_ctx = NULL;
 	}

--- a/src/hardware/hardware.cpp
+++ b/src/hardware/hardware.cpp
@@ -169,7 +169,6 @@ void ffmpeg_audio_frame_send() {
 
 	if (!pkt) E_Exit("Error: Unable to alloc packet");
 
-	ffmpeg_aud_frame->key_frame = 1;
 	ffmpeg_aud_frame->pts = (int64_t)ffmpeg_audio_sample_counter;
 	r=avcodec_send_frame(ffmpeg_aud_ctx,ffmpeg_aud_frame);
 	if (r < 0 && r != AVERROR(EAGAIN))
@@ -1528,7 +1527,7 @@ skip_shot:
 
 				// encode it
 				ffmpeg_vid_frame->pts = (int64_t)capture.video.frames; // or else libx264 complains about non-monotonic timestamps
-				ffmpeg_vid_frame->key_frame = ((capture.video.frames % 15) == 0)?1:0;
+                av_opt_set_int(ffmpeg_vid_ctx->priv_data, "g", 15, 0); // GOP size 15
 
 				r=avcodec_send_frame(ffmpeg_vid_ctx,ffmpeg_vid_frame);
 				if (r < 0 && r != AVERROR(EAGAIN))

--- a/src/hardware/hardware.cpp
+++ b/src/hardware/hardware.cpp
@@ -1764,7 +1764,7 @@ skip_mt_wav:
 }
 
 #pragma pack(push,1)
-typedef struct pcap_hdr_struct_t {
+typedef struct {
 	uint32_t magic_number;   /* magic number */
 	uint16_t version_major;  /* major version number */
 	uint16_t version_minor;  /* minor version number */
@@ -1772,14 +1772,14 @@ typedef struct pcap_hdr_struct_t {
 	uint32_t sigfigs;        /* accuracy of timestamps */
 	uint32_t snaplen;        /* max length of captured packets, in octets */
 	uint32_t network;        /* data link type */
-};
+} pcap_hdr_struct_t;
 
-typedef struct pcaprec_hdr_struct_t {
+typedef struct  {
 	uint32_t ts_sec;         /* timestamp seconds */
 	uint32_t ts_usec;        /* timestamp microseconds */
 	uint32_t incl_len;       /* number of octets of packet saved in file */
 	uint32_t orig_len;       /* actual length of packet */
-};
+} pcaprec_hdr_struct_t;
 #pragma pack(pop)
 
 void Capture_WritePacket(bool /*send*/,const unsigned char *buf,size_t len) {

--- a/src/hardware/hardware.cpp
+++ b/src/hardware/hardware.cpp
@@ -58,6 +58,7 @@ extern "C" {
 #include <libavcodec/avcodec.h>
 #include <libavformat/avformat.h>
 #include <libavformat/avio.h>
+#include <libavutil/opt.h>
 }
 
 /* This code now requires FFMPEG 4 or higher */

